File: app/src/main/java/com/expansion/lg/kimaru/expansion/tables/WardTable.java
package com.expansion.lg.kimaru.expansion.tables;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.util.Log;

import com.expansion.lg.kimaru.expansion.mzigos.SubCounty;
import com.expansion.lg.kimaru.expansion.mzigos.Ward;
import com.expansion.lg.kimaru.expansion.other.Constants;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by kimaru on 7/21/17.
 */

public class WardTable extends SQLiteOpenHelper {
    public static final String TABLE_NAME="wards";
    public static final String DATABASE_NAME= Constants.DATABASE_NAME;
    public static final int DATABASE_VERSION= Constants.DATABASE_VERSION;


    public static String varchar_field = " varchar(512) ";
    public static String primary_field = " _id INTEGER PRIMARY KEY AUTOINCREMENT ";
    public static String integer_field = " integer default 0 ";
    public static String text_field = " text ";

    public static String JSON_ROOT = "wards";

    public static final String ID = "id";
    public static final String NAME = "name";
    public static final String COUNTY = "county";
    public static final String SUBCOUNTY = "sub_county";
    public static final String ARCHIVED = "archived";

    String [] columns=new String[]{ID, NAME, COUNTY, SUBCOUNTY, ARCHIVED};

    public static final String CREATE_DATABASE="CREATE TABLE " + TABLE_NAME + "("
            + ID + varchar_field +", "
            + NAME + varchar_field + ", "
            + COUNTY + integer_field + ", "
            + SUBCOUNTY + varchar_field + ", "
            + ARCHIVED + integer_field + "); ";

    public static final String DATABASE_DROP="DROP TABLE IF EXISTS" + TABLE_NAME;

    public WardTable(Context context) {
        super(context, TABLE_NAME, null, DATABASE_VERSION);
    }
    @Override
    public void onCreate(SQLiteDatabase db) {

        db.execSQL(CREATE_DATABASE);

    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {

        Log.w("Link Facility", "upgrading database from" + oldVersion + "to" + newVersion);
        if (oldVersion < 2){
            upgradeVersion2(db);
        }
        if (oldVersion < 3){
            upgradeVersion3(db);
        }
    }


    public long addData(Ward ward) {

        SQLiteDatabase db=getWritableDatabase();
        ContentValues cv=new ContentValues();
        cv.put(ID, ward.getId());
        cv.put(NAME, ward.getName());
        cv.put(COUNTY, ward.getCounty());
        cv.put(SUBCOUNTY, ward.getSubCounty());
        cv.put(ARCHIVED, ward.getArchived());

        long id;
        if (isExist(ward)){
            id = db.update(TABLE_NAME, cv, ID+"='"+ward.getId()+"'", null);
            Log.d("Tremap DB Op", "Ward updated");
        }else{
            id = db.insertWithOnConflict(TABLE_NAME, null, cv, SQLiteDatabase.CONFLICT_REPLACE);
            Log.d("Tremap DB Op", "Ward Created");
        }
        db.close();
        return id;

    }
    public boolean isExist(Ward ward) {
        SQLiteDatabase db = getReadableDatabase();
        Cursor cur = db.rawQuery("SELECT id FROM " + TABLE_NAME + " WHERE "+ID+" = '" + ward.getId() + "'", null);
        boolean exist = (cur.getCount() > 0);
        cur.close();
        return exist;
    }

    public List<Ward> getAllWardsData() {

        SQLiteDatabase db=getReadableDatabase();

        Cursor cursor=db.query(TABLE_NAME,columns,null,null,null,null,null,null);

        List<Ward> wardList =new ArrayList<>();
        for (cursor.moveToFirst(); !cursor.isAfterLast();cursor.moveToNext()){
            Ward ward =new Ward();

            ward.setId(cursor.getString(cursor.getColumnIndex(ID)));
            ward.setName(cursor.getString(cursor.getColumnIndex(NAME)));
            ward.setCounty(cursor.getInt(cursor.getColumnIndex(COUNTY)));
            ward.setSubCounty(cursor.getString(cursor.getColumnIndex(SUBCOUNTY)));
            ward.setArchived(cursor.getInt(cursor.getColumnIndex(ARCHIVED)));
            wardList.add(ward);
        }
        db.close();
        return wardList;
    }

    public List<Ward> getWardsbySubCounty(SubCounty subCounty) {

        SQLiteDatabase db=getReadableDatabase();

        String whereClause = SUBCOUNTY+" = ? ";

        String[] whereArgs = new String[] {
                subCounty.getId()
        };
        Cursor cursor=db.query(TABLE_NAME,columns,whereClause,whereArgs,null,null,null,null);

        List<Ward> wardList =new ArrayList<>();
        for (cursor.moveToFirst(); !cursor.isAfterLast();cursor.moveToNext()){
            Ward ward =new Ward();
            ward.setId(cursor.getString(cursor.getColumnIndex(ID)));
            ward.setName(cursor.getString(cursor.getColumnIndex(NAME)));
            ward.setCounty(cursor.getInt(cursor.getColumnIndex(COUNTY)));
            ward.setSubCounty(cursor.getString(cursor.getColumnIndex(SUBCOUNTY)));
            ward.setArchived(cursor.getInt(cursor.getColumnIndex(ARCHIVED)));
            wardList.add(ward);
        }
        db.close();
        return wardList;
    }

    public List<Ward> getWardsbyCounty(Integer countyId) {

        SQLiteDatabase db=getReadableDatabase();

        String whereClause = COUNTY+" = ? ";

        String[] whereArgs = new String[] {
                countyId.toString()
        };
        Cursor cursor=db.query(TABLE_NAME,columns,whereClause,whereArgs,null,null,null,null);

        List<Ward> wardList =new ArrayList<>();
        for (cursor.moveToFirst(); !cursor.isAfterLast();cursor.moveToNext()){
            Ward ward =new Ward();
            ward.setId(cursor.getString(cursor.getColumnIndex(ID)));
            ward.setName(cursor.getString(cursor.getColumnIndex(NAME)));
            ward.setCounty(cursor.getInt(cursor.getColumnIndex(COUNTY)));
            ward.setSubCounty(cursor.getString(cursor.getColumnIndex(SUBCOUNTY)));
            ward.setArchived(cursor.getInt(cursor.getColumnIndex(ARCHIVED)));
            wardList.add(ward);
        }
        db.close();
        return wardList;
    }

    public Ward getWardById(String uuid) {

        SQLiteDatabase db=getReadableDatabase();

        String whereClause = ID+" = ? ";

        String[] whereArgs = new String[] {
                uuid
        };

        Cursor cursor=db.query(TABLE_NAME,columns,whereClause,whereArgs,null,null,null,null);
        if (!(cursor.moveToFirst()) || cursor.getCount() ==0){
            return null;
        }else {
            Ward ward =new Ward();
            ward.setId(cursor.getString(cursor.getColumnIndex(ID)));
            ward.setName(cursor.getString(cursor.getColumnIndex(NAME)));
            ward.setCounty(cursor.getInt(cursor.getColumnIndex(COUNTY)));
            ward.setSubCounty(cursor.getString(cursor.getColumnIndex(SUBCOUNTY)));
            ward.setArchived(cursor.getInt(cursor.getColumnIndex(ARCHIVED)));
            db.close();
            return ward;
        }
    }
    public void fromJson(JSONObject jsonObject){
        Ward ward = new Ward();
        try {

            ward.setId(jsonObject.getString(ID));
            String name= jsonObject.getString(NAME);
            if (!jsonObject.getString(NAME).equalsIgnoreCase("")){
                name = name.substring(0, 1).toUpperCase() + name.substring(1);
            }
            ward.setName(name);
            ward.setCounty(jsonObject.getInt(COUNTY));
            ward.setSubCounty(jsonObject.getString(SUBCOUNTY));
            ward.setArchived(jsonObject.getInt(ARCHIVED));
            addData(ward);
        }catch (Exception e){
            Log.d("Tremap ERR", "Ward from JSON "+e.getMessage());
        }
    }
    public JSONObject toJson(Cursor cursor) {
        JSONObject results = new JSONObject();

        JSONArray resultSet = new JSONArray();

        for (cursor.moveToFirst(); !cursor.isAfterLast();cursor.moveToNext()){
            int totalColumns = cursor.getColumnCount();
            JSONObject rowObject = new JSONObject();

            for (int i =0; i < totalColumns; i++){
                if (cursor.getColumnName(i) != null){
                    try {
                        if (cursor.getString(i) != null){
                            rowObject.put(cursor.getColumnName(i), cursor.getString(i));
                        }else{
                            rowObject.put(cursor.getColumnName(i), "");
                        }
                    }catch (Exception e){
                    }
                }
            }
            resultSet.put(rowObject);
            try {
                results.put(JSON_ROOT, resultSet);
            } catch (JSONException e) {

            }
        }
        cursor.close();
        return results;
    }

    public JSONObject getJson() {

        SQLiteDatabase db=getReadableDatabase();

        Cursor cursor=db.query(TABLE_NAME,columns,null,null,null,null,null,null);

        JSONObject results = new JSONObject();

        JSONArray resultSet = new JSONArray();

        for (cursor.moveToFirst(); !cursor.isAfterLast();cursor.moveToNext()){
            int totalColumns = cursor.getColumnCount();
            JSONObject rowObject = new JSONObject();

            for (int i =0; i < totalColumns; i++){
                if (cursor.getColumnName(i) != null){
                    try {
                        if (cursor.getString(i) != null){
                            rowObject.put(cursor.getColumnName(i), cursor.getString(i));
                        }else{
                            rowObject.put(cursor.getColumnName(i), "");
                        }
                    }catch (Exception e){
                    }
                }
            }
            resultSet.put(rowObject);
            try {
                results.put(JSON_ROOT, resultSet);
            } catch (JSONException e) {

            }
        }
        cursor.close();
        db.close();
        return results;
    }


    public void upgradeVersion2(SQLiteDatabase db){}
    public void upgradeVersion3(SQLiteDatabase db){}
}

File: app/src/main/java/com/expansion/lg/kimaru/expansion/tables/WardTable.kt
package com.expansion.lg.kimaru.expansion.tables

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.util.Log

import com.expansion.lg.kimaru.expansion.mzigos.SubCounty
import com.expansion.lg.kimaru.expansion.mzigos.Ward
import com.expansion.lg.kimaru.expansion.other.Constants

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

import java.util.ArrayList

/**
 * Created by kimaru on 7/21/17.
 */

class WardTable(context: Context) : SQLiteOpenHelper(context, TABLE_NAME, null, DATABASE_VERSION) {

    internal var columns = arrayOf(ID, NAME, COUNTY, SUBCOUNTY, ARCHIVED)

    val allWardsData: List<Ward>
        get() {

            val db = readableDatabase

            val cursor = db.query(TABLE_NAME, columns, null, null, null, null, null, null)

            val wardList = ArrayList<Ward>()
            cursor.moveToFirst()
            while (!cursor.isAfterLast) {
                val ward = Ward()

                ward.id = cursor.getString(cursor.getColumnIndex(ID))
                ward.name = cursor.getString(cursor.getColumnIndex(NAME))
                ward.county = cursor.getInt(cursor.getColumnIndex(COUNTY))
                ward.subCounty = cursor.getString(cursor.getColumnIndex(SUBCOUNTY))
                ward.archived = cursor.getInt(cursor.getColumnIndex(ARCHIVED))
                wardList.add(ward)
                cursor.moveToNext()
            }
            db.close()
            return wardList
        }

    val json: JSONObject
        get() {

            val db = readableDatabase

            val cursor = db.query(TABLE_NAME, columns, null, null, null, null, null, null)

            val results = JSONObject()

            val resultSet = JSONArray()

            cursor.moveToFirst()
            while (!cursor.isAfterLast) {
                val totalColumns = cursor.columnCount
                val rowObject = JSONObject()

                for (i in 0 until totalColumns) {
                    if (cursor.getColumnName(i) != null) {
                        try {
                            if (cursor.getString(i) != null) {
                                rowObject.put(cursor.getColumnName(i), cursor.getString(i))
                            } else {
                                rowObject.put(cursor.getColumnName(i), "")
                            }
                        } catch (e: Exception) {
                        }

                    }
                }
                resultSet.put(rowObject)
                try {
                    results.put(JSON_ROOT, resultSet)
                } catch (e: JSONException) {

                }

                cursor.moveToNext()
            }
            cursor.close()
            db.close()
            return results
        }

    override fun onCreate(db: SQLiteDatabase) {

        db.execSQL(CREATE_DATABASE)

    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {

        Log.w("Link Facility", "upgrading database from" + oldVersion + "to" + newVersion)
        if (oldVersion < 2) {
            upgradeVersion2(db)
        }
        if (oldVersion < 3) {
            upgradeVersion3(db)
        }
    }


    fun addData(ward: Ward): Long {

        val db = writableDatabase
        val cv = ContentValues()
        cv.put(ID, ward.id)
        cv.put(NAME, ward.name)
        cv.put(COUNTY, ward.county)
        cv.put(SUBCOUNTY, ward.subCounty)
        cv.put(ARCHIVED, ward.archived)

        val id: Long
        if (isExist(ward)) {
            id = db.update(TABLE_NAME, cv, ID + "='" + ward.id + "'", null).toLong()
            Log.d("Tremap DB Op", "Ward updated")
        } else {
            id = db.insertWithOnConflict(TABLE_NAME, null, cv, SQLiteDatabase.CONFLICT_REPLACE)
            Log.d("Tremap DB Op", "Ward Created")
        }
        db.close()
        return id

    }

    fun isExist(ward: Ward): Boolean {
        val db = readableDatabase
        val cur = db.rawQuery("SELECT id FROM " + TABLE_NAME + " WHERE " + ID + " = '" + ward.id + "'", null)
        val exist = cur.count > 0
        cur.close()
        return exist
    }

    fun getWardsbySubCounty(subCounty: SubCounty): List<Ward> {

        val db = readableDatabase

        val whereClause = "$SUBCOUNTY = ? "

        val whereArgs = arrayOf(subCounty.id)
        val cursor = db.query(TABLE_NAME, columns, whereClause, whereArgs, null, null, null, null)

        val wardList = ArrayList<Ward>()
        cursor.moveToFirst()
        while (!cursor.isAfterLast) {
            val ward = Ward()
            ward.id = cursor.getString(cursor.getColumnIndex(ID))
            ward.name = cursor.getString(cursor.getColumnIndex(NAME))
            ward.county = cursor.getInt(cursor.getColumnIndex(COUNTY))
            ward.subCounty = cursor.getString(cursor.getColumnIndex(SUBCOUNTY))
            ward.archived = cursor.getInt(cursor.getColumnIndex(ARCHIVED))
            wardList.add(ward)
            cursor.moveToNext()
        }
        db.close()
        return wardList
    }

    fun getWardsbyCounty(countyId: Int?): List<Ward> {

        val db = readableDatabase

        val whereClause = "$COUNTY = ? "

        val whereArgs = arrayOf(countyId!!.toString())
        val cursor = db.query(TABLE_NAME, columns, whereClause, whereArgs, null, null, null, null)

        val wardList = ArrayList<Ward>()
        cursor.moveToFirst()
        while (!cursor.isAfterLast) {
            val ward = Ward()
            ward.id = cursor.getString(cursor.getColumnIndex(ID))
            ward.name = cursor.getString(cursor.getColumnIndex(NAME))
            ward.county = cursor.getInt(cursor.getColumnIndex(COUNTY))
            ward.subCounty = cursor.getString(cursor.getColumnIndex(SUBCOUNTY))
            ward.archived = cursor.getInt(cursor.getColumnIndex(ARCHIVED))
            wardList.add(ward)
            cursor.moveToNext()
        }
        db.close()
        return wardList
    }

    fun getWardById(uuid: String): Ward? {

        val db = readableDatabase

        val whereClause = "$ID = ? "

        val whereArgs = arrayOf(uuid)

        val cursor = db.query(TABLE_NAME, columns, whereClause, whereArgs, null, null, null, null)
        if (!cursor.moveToFirst() || cursor.count == 0) {
            return null
        } else {
            val ward = Ward()
            ward.id = cursor.getString(cursor.getColumnIndex(ID))
            ward.name = cursor.getString(cursor.getColumnIndex(NAME))
            ward.county = cursor.getInt(cursor.getColumnIndex(COUNTY))
            ward.subCounty = cursor.getString(cursor.getColumnIndex(SUBCOUNTY))
            ward.archived = cursor.getInt(cursor.getColumnIndex(ARCHIVED))
            db.close()
            return ward
        }
    }

    fun fromJson(jsonObject: JSONObject) {
        val ward = Ward()
        try {

            ward.id = jsonObject.getString(ID)
            var name = jsonObject.getString(NAME)
            if (!jsonObject.getString(NAME).equals("", ignoreCase = true)) {
                name = name.substring(0, 1).toUpperCase() + name.substring(1)
            }
            ward.name = name
            ward.county = jsonObject.getInt(COUNTY)
            ward.subCounty = jsonObject.getString(SUBCOUNTY)
            ward.archived = jsonObject.getInt(ARCHIVED)
            addData(ward)
        } catch (e: Exception) {
            Log.d("Tremap ERR", "Ward from JSON " + e.message)
        }

    }

    fun toJson(cursor: Cursor): JSONObject {
        val results = JSONObject()

        val resultSet = JSONArray()

        cursor.moveToFirst()
        while (!cursor.isAfterLast) {
            val totalColumns = cursor.columnCount
            val rowObject = JSONObject()

            for (i in 0 until totalColumns) {
                if (cursor.getColumnName(i) != null) {
                    try {
                        if (cursor.getString(i) != null) {
                            rowObject.put(cursor.getColumnName(i), cursor.getString(i))
                        } else {
                            rowObject.put(cursor.getColumnName(i), "")
                        }
                    } catch (e: Exception) {
                    }

                }
            }
            resultSet.put(rowObject)
            try {
                results.put(JSON_ROOT, resultSet)
            } catch (e: JSONException) {

            }

            cursor.moveToNext()
        }
        cursor.close()
        return results
    }


    fun upgradeVersion2(db: SQLiteDatabase) {}
    fun upgradeVersion3(db: SQLiteDatabase) {}

    companion object {
        val TABLE_NAME = "wards"
        val DATABASE_NAME = Constants.DATABASE_NAME
        val DATABASE_VERSION = Constants.DATABASE_VERSION


        var varchar_field = " varchar(512) "
        var primary_field = " _id INTEGER PRIMARY KEY AUTOINCREMENT "
        var integer_field = " integer default 0 "
        var text_field = " text "

        var JSON_ROOT = "wards"

        val ID = "id"
        val NAME = "name"
        val COUNTY = "county"
        val SUBCOUNTY = "sub_county"
        val ARCHIVED = "archived"

        val CREATE_DATABASE = ("CREATE TABLE " + TABLE_NAME + "("
                + ID + varchar_field + ", "
                + NAME + varchar_field + ", "
                + COUNTY + integer_field + ", "
                + SUBCOUNTY + varchar_field + ", "
                + ARCHIVED + integer_field + "); ")

        val DATABASE_DROP = "DROP TABLE IF EXISTS$TABLE_NAME"
    }
}
